Bind event api tweaks

Blasz opened the pull request:
There are two API changes here:

1. Binding events are now achieved via a prop instead of a param to the withAnalytics HOC. This makes it easier to bind events without creating new components.

2. Events can be created on certain prop callbacks now via a map passed as the 2nd param to the withAnalytics HOC, similar to the old withAnalytics implementation.

mblaszczyk-atlassian committed:
Flow type repo and change bindEventsToProps from HOC param to prop

mblaszczyk-atlassian committed:
Add way to create events within components via HOC param

mblaszczyk-atlassian committed:
Add example of using createEventMap with functions

mblaszczyk-atlassian committed:
Change withAnalytics API to composable higher order function

This is more idiomatic HOC structure that allows this HOC to be composed
with other higher order functions. The only downside is that the
createEventMap arg is not always required so it may look a bit awkward
to call withAnalytics()(Component).

mblaszczyk-atlassian committed:
Add flow config so vscode flow checking works

mblaszczyk-atlassian committed:
Try to fix some flow type errors

mblaszczyk-atlassian committed:
Change listener behaviour to swallow events when fired

Multiple listeners still work when on different channels

mblaszczyk-atlassian committed:
Fix bug in page boundary where it was causing Namespace to return multiple elements

Blasz commented:
I've added a couple more things, the commit messages above should hopefully explain what I've added.

jaredcrowe commented on src/modules/analytics/AnalyticsEvent.js:
```diff
@@ -1,20 +1,31 @@
 // @flow
 
+export type noop = (...args: any) => {};
```
should return `void` not `{}` right?

Blasz replied:
Yeah, I just got rid of it since it's a subset of the FireAnalyticsEvent type (and flow was therefore complaining about it).

jaredcrowe commented on src/modules/analytics/AnalyticsEvent.js:
```diff
@@ -1,20 +1,31 @@
 // @flow
 
+export type noop = (...args: any) => {};
+export type FireAnalyticsEvent = (event: AnalyticsEvent, channel: string) => void | noop;
```
this reads to me like it returns `void` or `noop` - are you sure that this syntax is correct?

jaredcrowe commented on src/modules/analytics/AnalyticsEvent.js:
```diff
@@ -1,20 +1,31 @@
+export type noop = (...args: any) => {};
+export type FireAnalyticsEvent = (event: AnalyticsEvent, channel: string) => void | noop;
+export type RaiseAnalyticsEvent = (event: AnalyticsEvent) => void;
+export type EventEnhancer = (payload: {}) => {} | {};
```
this also reads like it returns `{} | {}` - can you please confirm 🙂

jaredcrowe commented on src/modules/analytics/withAnalytics.js:
```diff
@@ -1,141 +1,208 @@
-        return;
+import AnalyticsEvent, { type RaiseAnalyticsEvent } from './AnalyticsEvent';
+
+const noop = (...args: any) => {};
```
`void`?

Blasz replied:
I've changed to `undefined`, can you return `void` from a function or is it a type only thing?

jaredcrowe replied:
oh derp, no sorry this is correct. my head was in type mode still.

jaredcrowe commented on src/modules/analytics/withAnalytics.js:
```diff
@@ -1,141 +1,208 @@
+              if (typeof eventCreator === 'string') {
+                this.createAnalyticsEvent(eventCreator).raise();
+              } else if (typeof eventCreator === 'function') {
+                eventCreator(this.createAnalyticsEvent, props).raise();
```
I would suggest that we allow them to return `null` from this function (so we'll have do a null check before we call `.raise`). It's probably an edge case, but they might want to fire an event directly from the callback prop without ever raising an event.

jaredcrowe commented on src/modules/analytics/withAnalytics.js:
```diff
@@ -1,141 +1,208 @@
+
+        const props = {
+          ...permittedProps,
+          ...patchedProps,
```
i don't think we should be merging `patchedProps` here, just `permittedProps` right?

Blasz replied:
Yeah, good catch.

jaredcrowe reviewed (changes requested):
just a few minor comments but this is looking great dude

mblaszczyk-atlassian committed:
Address review feedback

1. Allow function variant of create event map handler to not raise event
   by returning null/undefined. Can be used to fire events from the
   handler directly instead of raise.
2. Add example of button that fires an event from a create event map
   handler and also fires on an empty channel to demonstrate that option

jaredcrowe commented on src/modules/analytics/withAnalytics.js:
```diff
@@ -1,141 +1,208 @@
-        return;
+import AnalyticsEvent, { type RaiseAnalyticsEvent } from './AnalyticsEvent';
+
+const noop = (...args: any) => undefined;
```
just change this back (sorry) and this should be good to go imo

Blasz replied:
Hmm, why does it need to be changed back?

jaredcrowe replied:
yeah i suppose there's no difference between `=> undefined` and `=> {}` hey 🤷‍♂️
